File: 392-is-subsequence/392-is-subsequence.cpp
int dp[101][10001];
class Solution {
    int sub(string &s, string &t, int n, int m)
    {
        if(n==0 || m==0)
        {
            return 0;
        }
        
        if(dp[n][m]!=-1)
        {
            return dp[n][m];
        }
        
        if(s[n-1]==t[m-1])
        {
            return dp[n][m] = 1+sub(s, t, n-1, m-1);
        }
        else
        {
            return dp[n][m] = max(sub(s, t, n-1, m), sub(s, t, n, m-1));
        }
    }
public:
    bool isSubsequence(string s, string t) {
        int n=s.size();
        int m=t.size();
        memset(dp, -1, sizeof(dp));
        return sub(s, t, n, m)==n;
    }
};

File: 392-is-subsequence/392-is-subsequence.java
class Solution {
    public boolean isSubsequence(String s, String t) {
         int n = s.length() , m = t.length();
        int[][] dp = new int[n+1][m+1];
        for(int i = 1; i <= n;i++){
            for(int j = 1; j <= m;j++){
                if(s.charAt(i-1) == t.charAt(j-1)){
                    dp[i][j] = 1 + dp[i-1][j-1];
                }else{
                    dp[i][j] = Math.max(dp[i-1][j] , dp[i][j-1]);
                }
            }
        }
        return dp[n][m] == n;
    }
}